### PR TITLE
DWARF: Do not reorder locals

### PR DESCRIPTION
**Pull request by kripken**
The binary writer reorders locals unconditionally. I forgot about this, and so
when I made DWARF disable optimization passes that reorder, this was
left active.

Optimally the writer would not do this, and the ReorderLocals pass would.
But it looks like we need special logic for tuple locals anyhow, as they
expand into multiple locals, so some amount of local order changes seems
unavoidable, is that right @tlively ?

Test changes are mostly just lots of offsets, and can be ignored, but
the new test `test/passes/dwarf-local-order.*` shows the issue. It
prints `$foo` once, then after a roundtrip (showing no reordering), then
it strips the DWARF section and prints after another roundtrip (which
does show reordering).

### CODE REVIEW

**Comment by kripken** [comment]
Updated PR to not look for DWARF sections again in each function - do
so once up front in the place that calls the functions.

This also makes us avoid the Stack IR writer if DWARF is present, which
matches what we do with source maps. This doesn't prevent any known
bugs, but it's simpler this way and debugging + Stack IR opts is not an
important combination.

File: src/wasm/wasm-stack.cpp [hunk 1]
```diff
@@ -1780,8 +1781,25 @@ void BinaryInstWriter::mapLocalsAndEmitHeader() {
...
+    FindAll<TupleExtract> extracts(func->body);
+    if (!extracts.list.empty()) {
+      Fatal() << "DWARF + multivalue is not yet complete";
+    }
+    Index varStart = func->getVarIndexBase();
+    Index varEnd = varStart + func->getNumVars();
+    o << U32LEB(func->getNumVars());
+    for (Index i = varStart; i < varEnd; i++) {
+      mappedLocals[std::make_pair(i, 0)] = i;
+      o << U32LEB(1) << binaryType(func->getLocalType(i));
```
Comment (tlively) [inline src/wasm/wasm-stack.cpp:1800]: I suppose it's probably not worth coalescing runs of identical types if we are disabling so many optimizations anyways.

File: src/wasm/wasm-stack.cpp [hunk 2]
```diff
@@ -1780,8 +1781,25 @@ void BinaryInstWriter::mapLocalsAndEmitHeader() {
...
+    mappedLocals[std::make_pair(i, 0)] = i;
+  }
+  // Normally we map all locals of the same type into a range of adjacent
+  // addresses, which is more compact. However, if we need to keep DWARF valid,
+  // do not do any reordering at all - instead, do a trivial mapping that
+  // keeps everything unmoved.
+  if (DWARF) {
+    FindAll<TupleExtract> extracts(func->body);
+    if (!extracts.list.empty()) {
+      Fatal() << "DWARF + multivalue is not yet complete";
```
Comment (tlively) [inline src/wasm/wasm-stack.cpp:1793]: In a module that contains a multivalue function call, do we know what the DWARF will look like? I suspect that the DWARF will be messed up during parsing, well before we emit the binary.

Comment (kripken) [reply src/wasm/wasm-stack.cpp:1793]: Good question, I'm not sure. Sounds like something we'll need to handle.

**Pull request merged**